mqtt/mqtt-bridge/src/persist/storage/ring_buffer/mod.rs

```diff
@@ -0,0 +1,416 @@
+                let end = start + data_size;
+                let bytes = &mmap[start..end];
+                data = binary_deserialize::<Data>(bytes)?;
+            } else {
+                let start = read_index + block_size;
+                let end = start + data_size;
+                let split = end - self.max_file_size;
+                let mmap = self.mmap.lock();
+                let left = &mmap[start..self.max_file_size];
+                let right = &mmap[0..split];
+                let mut bytes = Vec::from(left);
+                bytes.append(&mut Vec::from(right));
+                data = binary_deserialize::<Data>(&bytes)?;
+            }
+            validate(&block, &data)?;
+            let something = binary_deserialize::<Vec<u8>>(data.inner())?;
+            vdata.push_back((index, something));
+            read_index += block_size + data_size;
+        }
+
+        self.pointers.read.store(
+            (read_index + total_size) % self.max_file_size,
+            Ordering::SeqCst,
+        );
+
+        Poll::Ready(Ok(vdata))
+    }
+}
+
+struct RingBufferRemoveFuture {
+    pointers: Arc<FilePointers>,
+    flush_options: FlushOptions,
+    mmap: Arc<Mutex<MmapMut>>,
+    flush_state: Arc<FlushState>,
+    key: usize,
+    max_file_size: usize,
+}
+
+impl Future for RingBufferRemoveFuture {
+    type Output = StorageResult<()>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let timer = Instant::now();
+        let write_index = self.pointers.write.load(Ordering::SeqCst);
+        let block_size_result = binary_serialize_size(&BlockHeaderWithHash::new(0, 0, 0));
+
+        if block_size_result.is_err() {
+            return Poll::Ready(Err(RingBufferError::Serialization(
+                block_size_result.unwrap_err(),
+            )
+            .into()));
+        }
+
+        let block_size = block_size_result.unwrap();
+
+        if self.key >= write_index {
+            return Poll::Pending;
+        }
+        cx.waker().wake_by_ref();
+
+        let start = self.key;
+        let end = start + block_size;
+        if end > self.max_file_size {
+            return Poll::Ready(Err(RingBufferError::WrapAround.into()));
+        }
+        let bytes = mmap_read(self.mmap.clone(), start, end);
+
+        let mut block = binary_deserialize::<BlockHeaderWithHash>(&bytes)?;
+
+        let inner_block = block.inner_mut();
+        inner_block.set_should_not_overwrite(false);
+
+        let bytes = binary_serialize(&block)?;
+
+        let should_flush = should_flush(self.flush_options, &self.flush_state);
+        mmap_write(self.mmap.clone(), start, end, &bytes, should_flush)?;
+
+        self.flush_state
+            .update(0, 0, timer.elapsed().as_millis() as usize);
+        if should_flush {
+            self.flush_state.reset(&self.flush_options);
+        }
+
+        Poll::Ready(Ok(()))
+    }
+}
+
+impl Storage<usize, Vec<u8>> for RingBuffer {
```

Comment:
One thing I thought about is how to guarantee we don't read some data and have it in memory but write over it on disk from input thinking everything is fine when in reality we might crash. I think the right thing to do is something like this:
1. Have each read flag the data as 'in progress' or something
2. On read completion of the data (most likely remove operation) update the flag to 'done'
3. Have a BatchResults struct that keep track of these
4. Once all are done the BatchResults can update the read pointer

If we were to crash that will be okay because we can come back up without having to worry about writing over read after we did a batch. It would also allow us to avoid duplicates after crashing. I still need to ponder this idea a bit... but will push after I have something that looks reasonable to discuss.

Reply:
Upon thinking on this more, I think I can just use two read pointers, one a begin and one an end. The begin one will be updated in "remove" and the end updated in "batch". The "remove" operation will continue to read ahead and check if there are valid entries after a removal. If there are valid entries that were marked for removal, then continue to move up the read begin pointer. In "batch" the end read pointer will be at the end of the last element in the batch.

Reply:
I like this idea.

Reply:
As discussed offline, we can have a single read pointer and error if the key for remove does not match the pointer. As we are only expecting to maintain in-order transfer as per MQTT